func! macro lifetime mismatch for generic functions

Hi,

I'm trying to mock `nix::unistd::execvp` which has the function signature below:

```
pub fn execvp<S: AsRef<CStr>>(
    filename: &CStr,
    args: &[S],
) -> Result<Infallible>
```

when I use InjectorPP with the lines below:

```
injector.when_called(injectorpp::func!(nix::unistd::execvp, fn(&std::ffi::CStr, &[&std::ffi::CStr]) -> nix::Result<std::convert::Infallible>))
        .will_execute(injectorpp::fake!(
            func_type: fn(_cmd: &std::ffi::CStr, _args: &[&std::ffi::CStr]) -> nix::Result<std::convert::Infallible>,
            returns: Err(nix::errno::Errno::E2BIG)
        )
);
```

I'm getting the compile time error:

```
error[E0308]: mismatched types
   --> src/linux_util/namespace/mod.rs:221:48
    |
221 |         injector.when_called(injectorpp::func!(nix::unistd::execvp, fn(&std::ffi::CStr, &[&std::ffi::CStr]) -> nix::Result<std::convert::Infallible>))
    |                                                ^^^^^^^^^^^^^^^^^^^  -------------------------------------------------------------------------------- expected due to this
    |                                                |
    |                                                one type is more general than the other
    |
    = note: expected fn pointer `for<'a, 'b, 'c> fn(&'a CStr, &'b [&'c CStr]) -> Result<_, _>`
                  found fn item `for<'a, 'b> fn(&'a CStr, &'b [_]) -> Result<_, _> {nix::unistd::execvp::<_>}`
```

My questions are:

* I couldn't manage to avoid the lifetime mismatch between my mock signature and the real `execvp` signature. I don't see a way to define the lifetimes for the parameters within the `func!` macro. Is there any ways to achieve this?
* I tried to use `&'static` lifetime for all the parameters in the mock signature to match the lifetime expectations and it helped me to avoid the compile time failure but the injector didn't catch the call to `execvp` at all.

Thanks!

--- a/src/interface/macros.rs
+++ b/src/interface/macros.rs
@@ -19,9 +19,8 @@
 macro_rules! func {
     // Case 1: Generic function — provide function name and types separately
     ($f:ident :: <$($gen:ty),*>, $fn_type:ty) => {{
-        let fn_val:$fn_type = $f::<$($gen),*>;
-        let ptr = fn_val as *const ();
-        let sig = std::any::type_name_of_val(&fn_val);
+        let ptr = $f::<$($gen),*> as *const ();
+        let sig = std::any::type_name::<$fn_type>();
         let type_id = std::any::TypeId::of::<$fn_type>();
 
         unsafe { FuncPtr::new_with_type_id(ptr, sig, type_id) }
@@ -29,9 +28,8 @@ macro_rules! func {
 
     // Case 2: Non-generic function with explicit type
     ($f:expr, $fn_type:ty) => {{
-        let fn_val:$fn_type = $f;
-        let ptr = fn_val as *const ();
-        let sig = std::any::type_name_of_val(&fn_val);
+        let ptr = ($f) as *const ();
+        let sig = std::any::type_name::<$fn_type>();
         let type_id = std::any::TypeId::of::<$fn_type>();
 
         unsafe { FuncPtr::new_with_type_id(ptr, sig, type_id) }

--- a/tests/will_execute.rs
+++ b/tests/will_execute.rs
@@ -572,3 +572,31 @@ fn test_will_execute_fake_unsafe_unit_assign_and_times_over_called_should_panic(
     });
     assert!(result.is_err());
 }
+
+// Generic function where type parameter S appears inside a slice reference.
+// When monomorphized with S = &str, the fn pointer type has different lifetime
+// counts than the explicit fn pointer type (issue #107).
+fn generic_with_ref_type_param<S: AsRef<str>>(
+    _prefix: &str,
+    _items: &[S],
+) -> String {
+    "original".to_string()
+}
+
+#[test]
+fn test_will_execute_generic_func_with_ref_type_param_via_turbofish() {
+    let mut injector = InjectorPP::new();
+    injector
+        .when_called(injectorpp::func!(
+            generic_with_ref_type_param::<&str>,
+            fn(&str, &[&str]) -> String
+        ))
+        .will_execute(injectorpp::fake!(
+            func_type: fn(_prefix: &str, _items: &[&str]) -> String,
+            returns: "mocked".to_string(),
+            times: 1
+        ));
+
+    let result = generic_with_ref_type_param("hello", &["a", "b"]);
+    assert_eq!(result, "mocked");
+}

